src/core/EdFi.DataManagementService.Core/Extraction/ReferenceExtractor.cs

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Diagnostics;
+using System.Text.Json.Nodes;
+using EdFi.DataManagementService.Core.ApiSchema;
+using EdFi.DataManagementService.Core.ApiSchema.Extensions;
+using EdFi.DataManagementService.Core.External.Model;
+using EdFi.DataManagementService.Core.Model;
+using Microsoft.Extensions.Logging;
+
+namespace EdFi.DataManagementService.Core.Extraction;
+
+/// <summary>
+/// Extracts document references for a resource
+/// </summary>
+internal static class ReferenceExtractor
+{
+    /// <summary>
+    /// Takes an API JSON body for the resource and extracts the document reference information from the JSON body.
+    /// </summary>
+    public static DocumentReference[] ExtractReferences(this ResourceSchema resourceSchema, JsonNode documentBody, ILogger _logger)
+    {
+        _logger.LogDebug("ReferenceExtractor.Extract");
+
+        List<DocumentReference> result = [];
+
+        foreach (DocumentPath documentPath in resourceSchema.DocumentPaths)
+        {
+            if (!documentPath.IsReference)
+                continue;
+            if (documentPath.IsDescriptor)
+                continue;
+
+            // Extract the reference values from the document
+            IntermediateReferenceElement[] intermediateReferenceElements = documentPath
+                .ReferenceJsonPathsElements.Select(
+                    referenceJsonPathsElement => new IntermediateReferenceElement(
+                        referenceJsonPathsElement.IdentityJsonPath,
+                        documentBody
+                            .SelectNodesFromArrayPathCoerceToStrings(
+                                referenceJsonPathsElement.ReferenceJsonPath.Value,
+                                _logger
+                            )
+                            .ToArray()
+                    )
+                )
+                .ToArray();
+
+            // If a JsonPath selection had no results, we can assume an optional reference wasn't there
+            if (Array.Exists(intermediateReferenceElements, x => x.ValueSlice.Length == 0))
+                continue;
+
+            int valueSliceLength = intermediateReferenceElements[0].ValueSlice.Length;
+
+            // Number of document values from resolved JsonPaths should all be the same, otherwise something is very wrong
+            Trace.Assert(
```

Comment:
Should this `Trace.Assert` happen before line 53? And then on that `if` statement instead of looking at each `intermediateReferenceElements` just test the first one for `ValueSlice.Length == 0`? Maybe I'm misunderstanding the possibilities here.